### DISCUSSION TITLE
Exception when add a dxf body importer

### DISCUSSION THREAD

**Discussion by RileyGe** [Macad3D/Macad3D#25] [Q&A] [answered]
I want to import body from a dxf file. To do this, I add an IBodyImporter interface to the Class DxfExchanger. The main functions of Importer are implemented in the DxfBodyImporter class.

Please find code from here: https://github.com/RileyGe/Macad3D/commit/5f6be7e99890625db02694bc6c0d462496d551ef

When I import the file(Please find the file in the attachment) P31-2.dxf, every thing work fine. But when I import P31-6.dxf, Exception happens whith information:
System.AccessViolationException:“Attempted to read or write protected memory. This is often an indication that other memory is corrupt.”

![image](https://user-images.githubusercontent.com/26666371/190079381-8ddc7244-ae8f-4f2d-9482-b7445c16ec28.png)

![image](https://user-images.githubusercontent.com/26666371/190079191-3c003510-adbe-4f8c-8916-5fe76a408af7.png)

I try my best to find the error, I found when adding the third face exception happens. But I have little experience of occ coding, so I have trouble to fix this.

![Snipaste_2022-09-14_14-08-37](https://user-images.githubusercontent.com/26666371/190080077-f0d67ae3-9787-4643-8b74-6c8739ffb621.png)

Thanks

[file.zip](https://github.com/Macad3D/Macad3D/files/9563403/file.zip)

**Comment by dipts** [comment, marked as answer]
This probably means that one of the two shapes is not valid. You can activate "enable native debugger" in the project settings to see what is happening in the OCCT and where something is going wrong.

Is one of the native instance pointers null?

**Reply by RileyGe** [reply]
You are right, after I enabled the native debugger, the "myTShape" of the second instance is null. Please find the picture below for details.

But still I can not find a mothed to solve the issue, help please.

![image](https://user-images.githubusercontent.com/26666371/190570823-597a6f9e-db55-4bb2-b979-3946ed2d93f8.png)

![image](https://user-images.githubusercontent.com/26666371/190570909-41d6c039-6a9e-45dd-82e7-fa3aa0e41686.png)

![image](https://user-images.githubusercontent.com/26666371/190570368-088d665e-e6e3-4841-a9f2-83aff964b006.png)

More details, the four points which to build the face is in the same plane. But the three faces cannot build a convex polyhedron. See the picture below.

![image](https://user-images.githubusercontent.com/26666371/190571943-ed0a17b5-d245-402a-aebf-b2898eac252e.png)

**Reply by dipts** [reply]
The shapes do not contain any geometry, so for some reason the MakeFace algorithm cannot build the face. Maybe an precision problem.

I'm not quite sure what you're importing. When you read the FACE entities from a DXF, you get quads and triangles - more of a mesh than a solid BRep shape. You can put the triangles into a triangulation and import it as a mesh shape, this avoids many problems with the replication. 

If it is to be editable as a BRep, the shape must be closed and watertight at the end. OCCT has some algorithms to ensure this (keyword Shape Healing) that can help. Getting the roundings round again is more difficult. Is there no way to get the shape as a real BRep, e.g. as IGES or STEP?

**Reply by RileyGe** [reply]
You are right on the issue, I finally find out where is the error. The four points is not in the same plane. So I update the `MakeFace` method like this:
```
public static TopoDS_Face MakeFace(Pnt[] points)
{
    var edges = new List<TopoDS_Edge>();
    for (int i = 0; i < points.Length; i++)
    {
        edges.Add(new BRepBuilderAPI_MakeEdge(points[i], points[i == points.Length - 1 ? 0 : i + 1]).Edge());
    } 
    var wire = points.Length == 3 ? new BRepBuilderAPI_MakeWire(edges[0], edges[1], edges[2]).Wire() :
        new BRepBuilderAPI_MakeWire(edges[0], edges[1], edges[2], edges[3]).Wire();
    var face = new BRepBuilderAPI_MakeFace(wire, true);
    if(face.Error() != BRepBuilderAPI_FaceError.BRepBuilderAPI_FaceDone)
    {
        switch (face.Error())
        {
            case BRepBuilderAPI_FaceError.BRepBuilderAPI_NotPlanar: 
                Dir v1 = (points[1] - points[0]).ToDir();
                Dir v2 = (points[2] - points[0]).ToDir();
                Pln pl = new Pln(points[0], v1.Crossed(v2));
                var pt2d = ProjLib.Project(pl, points[3]);
                var pt4 = ElSLib.Value(pt2d.X, pt2d.Y, pl);
                points[^1] = pt4;
                return MakeFace(points);
        }
    }
    return face.Face();
}
```
This is not elegant, but it works!

Besides, I am working on how to set the tolerance of opencascade, I think this way may be more elegant than update the `MakeFace` method.

**Reply by RileyGe** [reply]
Finally, I solved the problem. If you want to set the tolerance when building a face, you should first set the tolerance of the edges.
You can find the example below. At line 6~8, I set the tolerance of the edge, then everything will be ok.
```
public static TopoDS_Face MakeFace(Pnt[] points)
{
    var edges = new List<TopoDS_Edge>();
    for (int i = 0; i < points.Length; i++)
    {
        var edge = new BRepBuilderAPI_MakeEdge(points[i], points[i == points.Length - 1 ? 0 : i + 1]).Edge();
        BRep_Builder edgeBuilder = new();
        edgeBuilder.UpdateEdge(edge, 0.001);
        edges.Add(edge);
    } 
    var wire = points.Length == 3 ? new BRepBuilderAPI_MakeWire(edges[0], edges[1], edges[2]).Wire() :
        new BRepBuilderAPI_MakeWire(edges[0], edges[1], edges[2], edges[3]).Wire();
    return new BRepBuilderAPI_MakeFace(wire).Face();
}
```

**Reply by dipts** [reply]
This is good news. Thank you for sharing your insights.

**Comment by xyont** [comment]
hi, this such a great feature in conversion DXF to STEP files. Is the code is merged and available or supported officially?
